bser_parse.c

```diff
@@ -0,0 +1,311 @@
+
+#include "bser.h"
+
+#include <stdio.h>
+#include <string.h>
+#include <jansson.h>
+
+static bser_t* bser_new_unparsed(
+        bser_buffer_t* buffer, bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_UNPARSED;
+    fill->value.unparsed.buffer = buffer;
+    return fill;
+}
+
+static bser_t* bser_new_error(
+        const char* message, bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_ERROR;
+    fill->value.error_message = message;
+    return fill;
+}
+
+static bser_t* bser_new_compact_object(
+        bser_buffer_t* buffer, bser_t* header, bser_t* fill) {
+    size_t sz = bser_array_size(header);
+    bser_key_value_pair_t* fields = (bser_key_value_pair_t*)malloc(
+        sizeof(bser_key_value_pair_t) * sz);
+
+    for (int i = 0; i < sz; ++i) {
+        fields[i].key = *(bser_array_get(header, i));
+        bser_new_unparsed(buffer, &fields[i].value);
+    }
+    bser_new_object(fields, sz, fill);
+    return fill;
+}
+
+bser_t* bser_parse(uint8_t* data, size_t buflen, bser_t* fill) {
+    if (fill == NULL) {
+        fill = (bser_t*)malloc(sizeof(bser_t));
```

Comment:
no need to cast the result of malloc -- it's a void* which can be assigned to anything.